src/Npgsql/PoolManager.cs

```diff
@@ -103,6 +103,9 @@ internal static void ClearAll()
                         return;
                     cp.Pool?.Clear();
                 }
+
+                _nextSlot = 0;
```

Comment:
I think there's a rare race condition here with TryGetValue:

1. _pools is bigger than InitialPoolsSize (has already been extended).
2. TryGetValue starts running, reads _nextSlot but not yet _pools.
3. Someone calls ClearAllPools, _pools is replaced with a new, smaller array (back to InitialPoolsSize)
4. TryGetValue resumes, reads the new _pools and tries to iterate beyond the size of the array.

Assuming the above is true, this can be avoided by doing a Min between nextSlot and pools.Length after both of these have been read.

Reply:
Yep, there is one. Fixed.